internal/clients/http/rbac/rbac_client.go

```diff
@@ -0,0 +1,119 @@
+package rbac
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients"
+	"github.com/RHEnVision/provisioning-backend/internal/clients/http"
+	"github.com/RHEnVision/provisioning-backend/internal/config"
+	"github.com/RHEnVision/provisioning-backend/internal/headers"
+	"github.com/RHEnVision/provisioning-backend/internal/metrics"
+	"github.com/RHEnVision/provisioning-backend/internal/ptr"
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/otel"
+)
+
+const TraceName = "github.com/EnVision/provisioning/internal/clients/http/rbac"
+
+type rbac struct {
+	client *ClientWithResponses
+}
+
+func init() {
+	clients.GetRbacClient = newClient
+}
+
+func logger(ctx context.Context) zerolog.Logger {
+	return zerolog.Ctx(ctx).With().Str("client", "ib").Logger()
+}
+
+func newClient(ctx context.Context) clients.Rbac {
+	if !config.Application.RbacEnabled {
+		return &allPermRbacSingleton
+	}
+
+	c, err := NewClientWithResponses(config.RBAC.URL, func(c *Client) error {
+		c.Client = http.NewPlatformClient(ctx, config.ImageBuilder.Proxy.URL)
+		return nil
+	})
+	if err != nil {
+		// in case there was an error during initialization return "no permissions ACL"
+		zerolog.Ctx(ctx).Warn().Err(err).Msg("Could not initialize RBAC client, returning empty ACL")
+		return &noPermRbacSingleton
+	}
+	return &rbac{client: c}
+}
+
+func (c *rbac) Ready(ctx context.Context) error {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "Ready")
+	defer span.End()
+
+	logger := logger(ctx)
+	resp, err := c.client.GetStatus(ctx, headers.AddRbacIdentityHeader, headers.AddEdgeRequestIdHeader)
+	if err != nil {
+		logger.Error().Err(err).Msgf("Readiness request failed for RBAC: %s", err.Error())
+		return err
+	}
+	defer resp.Body.Close()
+
+	err = http.HandleHTTPResponses(ctx, resp.StatusCode)
+	if err != nil {
+		return fmt.Errorf("ready call: %w", err)
+	}
+	return nil
+}
+
+// ErrMetaNotPresent is returned when metadata for pagination is not present in the response.
+var ErrMetaNotPresent = fmt.Errorf("RBAC did not return metadata: %w", clients.HttpClientErr)
+
+// Maximum possible entries returned in one request
+var FetchLimit = ptr.To(QueryLimit(500))
+
+func (c *rbac) GetPrincipalAccess(ctx context.Context) (clients.RbacAcl, error) {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "GetPrincipalAccess")
+	defer span.End()
+
+	start := time.Now()
+	defer func() {
+		metrics.RbacAclFetchDuration.Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
+	}()
+
+	var result clients.AccessList
+	var records int64 = math.MaxInt64
+
+	// keep fetching until we have all the records
+	for int64(len(result)) < records {
```

Comment:
This condition is kinda weird, can you explain? I think this will loop forever

Reply:
You are right there was a bug, I forgot to add offset so it actually stacked the first N elements to the slice over and over again. It did not loop tho, fixed. I tested it by adding some roles to my account so it returned a longer list and then I set the FetchLimit to a small number.

Reply:
You know what, there is a possibility for an endless loop which is really bad - if there is a bug in RBAC service and it returns "total" number that is higher than actual records and in subsequent queries it will return empty array. No results will be ever appended and it will continue looping. And it is possible to get here in case someone deletes a permission just after you receive the first page.

So I added maximum limit of queries (10), if that ever happens a warning is logged and the result is returned.

Consequence: on an account with more than 5000 permissions, some might be ignored. Meaning access denied.